### ISSUE TITLE
Remove dependency on lens

### ISSUE THREAD

**Issue by tonymorris** [NICTA/validation#1]
It would be nice to remove the dependency on the `lens` package and use smallerpackages such as `profunctors`. Unfortunately, this means foregoing the `Swapped` instances.

Discussion on this issue is here:
https://github.com/ekmett/lens/issues/526

**Comment by gwils** [comment]
I don't think the benefit of removing the lens dependency outweighs the usefulness of the Swapped instance.

**Comment by recursion-ninja** [comment]
Since Cabal 3.0 allows for multiple libraries to be exported from a package, we could split the dependencies up into two exposed libraries.

`validation` would not depend on `lens` and forgo the `Swapped` instance.
`validation:lens` would depend on `lens` and additionally expose a `Swapped` instance.

Users of the Library would have to import two modules to use the `Swapped` instance, say something like:
```
Data.Validation
Data.Validation.Lens
```

where the first module is exposed by `validation` and the second module is exposed from `validation:lens`.

The advantage of this approach is that if a user want to use the `Validation` type but does not need the `Swapped` instance, they can have their project depend on only the `validation` package and not also the `validation:lens` package, which allows their project to avoid the transitive dependency on `lens`.

**Comment by gwils** [comment]
Unfortunately that would make `instance Swapped Validation` an orphan instance (as would having it in a separate cabal package).

**Comment by recursion-ninja** [comment]
You're not wrong. Would that really be a problem since all instances for the `Validation` type originate from libraries of the `validation` package? Philosophically, it seems fine to me.

There's always some trade-offs to be made. I think the orphan instance(s) are a minor triviality and the convenience for the library consumer is major win. I greatly wish more packages did this so that I could opt out of depending on `lens` and `aeson` when my project uses neither of them.

**Comment by tonymorris** [comment]
I think the penalty of orphan instances is very high and to be avoided. If something as trivial as depending on lens resolves this penalty, then I think it is a reasonable solution.

**Comment by kindaro** [comment]
@tonymorris A dependency on `lens` is not trivial. `lens` is a dividing line between two _«ecosystems»_, two coding styles, and two vastly separated levels of expertise. `Validation` as an idea is simple, clear, _«mathematical»_ and, so to say, _«should be in `base`»_. But any number of people will avoid it just to stay away from the _«kmettosphere»_. Not to mention crazy build time. So, the minor convenience of `Swapped` without orphan instances should be weighed against people and projects being excluded from the user base.

**Comment by kindaro** [comment]
I took the liberty of making a [fork][fork] without any `lens` stuff. I will see if it works for me. I would prefer a more moderate edit, but this is the best I have time for.

[fork]: https://github.com/kindaro/validation/tree/e18fb22a49e21c9ad6cfbbe20c363a26b1bd09e0

**Comment by aloussase** [comment]
Couldn't the lens dependency be moves to another package that uses the same modules? So that users can opt-in for the Swappable instance.
